packages/wrangler/src/__tests__/configuration.test.ts

```diff
@@ -1364,6 +1364,78 @@ describe("normalizeAndValidateConfig()", () => {
       expect(diagnostics.hasWarnings()).toBe(false);
     });
 
+    describe("non-legacy", () => {
+      it("should use top-level `name` field", () => {
+        const rawConfig: RawConfig = {
+          name: "NAME",
+          legacy_env: false,
+          env: { DEV: {} },
+        };
+
+        const { config, diagnostics } = normalizeAndValidateConfig(
+          rawConfig,
+          undefined
+        );
+
+        expect(config.env.DEV.name).toEqual("NAME");
+        expect(diagnostics.hasErrors()).toBe(false);
+        expect(diagnostics.hasWarnings()).toBe(false);
+      });
+
+      it("should error if named environment contains a `name` field, even if there is no top-level name", () => {
+        const rawConfig: RawConfig = {
+          legacy_env: false,
+          env: {
+            DEV: {
+              name: "ENV_NAME",
+            },
+          },
+        };
+
+        const { config, diagnostics } = normalizeAndValidateConfig(
+          rawConfig,
+          undefined
+        );
+
+        expect(config.env.DEV.name).toBeUndefined();
+        expect(diagnostics.renderErrors()).toMatchInlineSnapshot(`
+          "Processing wrangler configuration:
+
+            - \\"env.DEV\\" environment configuration
+              - The \\"name\\" field is not allowed in non-legacy named environments.
```

Comment:
We shouldn't use the "non-legacy" phrasing here I think, "named service environments" maybe?